auraed/src/ebpf/tracepoint_programs/tracepoint_program.rs

```diff
@@ -0,0 +1,130 @@
+use super::PerfEventListener;
+use aya::programs::ProgramError;
+use aya::{
+    maps::perf::AsyncPerfEventArray,
+    programs::TracePoint,
+    util::{nr_cpus, online_cpus},
+    Bpf,
+};
+use bytes::BytesMut;
+use log::trace;
+use procfs::page_size;
+use std::mem::size_of;
+use tokio::sync::broadcast;
+use tracing::error;
+
+/// Size (in pages) for the circular per-CPU buffers that BPF perfbuf creates.
+const PER_CPU_BUFFER_SIZE_IN_PAGES: usize = 2;
+
+pub trait TracepointProgram<T: Clone + Send + 'static> {
+    const PROGRAM_NAME: &'static str;
+    const CATEGORY: &'static str;
+    const EVENT: &'static str;
+    const PERF_BUFFER: &'static str;
+
+    fn load_and_attach(
+        bpf: &mut Bpf,
+    ) -> Result<PerfEventListener<T>, anyhow::Error> {
+        trace!("Loading eBPF program: {}", Self::PROGRAM_NAME);
+
+        // Load the eBPF TracePoint program
+        let program: &mut TracePoint = bpf
+            .program_mut(Self::PROGRAM_NAME)
+            .ok_or_else(|| anyhow::anyhow!("failed to get eBPF program"))?
+            .try_into()?;
+
+        // Load the program
+        match program.load() {
+            Ok(_) => Ok(()),
+            Err(ProgramError::AlreadyLoaded) => Ok(()),
+            other => other,
+        }?;
+
+        // Query the number of CPUs on the host
+        let num_cpus = nr_cpus()?;
+
+        // Query the page size on the host
+        let page_size = page_size()?;
+
+        // Get the size of the event payload
+        let event_struct_size: usize = size_of::<T>();
+
+        // Calculate the capacity of the per-CPU buffers based on the size of
+        // the event
+        let per_cpu_buffer_capacity = (PER_CPU_BUFFER_SIZE_IN_PAGES
+            * page_size as usize)
+            / event_struct_size;
+
+        // Set the capacity of the channel to the combined capacity of all the
+        // per-CPU buffers
+        let channel_capacity = per_cpu_buffer_capacity * num_cpus;
+
+        // Attach to kernel trace event
+        let _ = program.attach(Self::CATEGORY, Self::EVENT)?;
```

Comment:
I didn't ignore the `AlreadyAttached` error that this can throw. Should I?

Reply:
Same as above ☝️